lib/write-set-tokens.ts

```diff
@@ -0,0 +1,90 @@
+import { types, Set, Range, Char, SetTokens } from './types';
+import * as sets from './sets-lookup';
+
+/**
+ * Takes character code and returns character to be displayed in a set
+ * @param {number} charCode Character code of set element
+ * @param {boolean} isFirstChar True if character is the first character of the set
+ * @param {boolean} isLastChar True if character is the last character of the set
+ * @returns {string} The string for the sets character
+ */
+export function setChar(charCode: number, isFirstChar: boolean, isLastChar = false): string {
+  // We only need to negate first internal '^' if token.not is false
+  return isFirstChar && charCode === 94 ? '\\^' :
+    charCode === 92 ? '\\\\' :
+      charCode === 93 ? '\\]' :
+        charCode === 45 && !isLastChar ? '\\-' :
+          String.fromCharCode(charCode);
+}
+
+/**
+ * Test if a character set matches a 'set-lookup'
+ * @param {SetTokens} set The set to be tested
+ * @param {Record<string | number, boolean>} map The predefined 'set-lookup'
+ * @returns {boolean} True if the character set corresponds to the 'set-lookup'
+ */
+function isSameSet(set: SetTokens, map: Record<string | number, boolean>): boolean {
+  for (const elem of set) {
+    if (elem.type === types.SET) {
+      return false;
+    }
+    const key = elem.type === types.CHAR ? elem.value : `${elem.from}-${elem.to}`;
+    if (map[key]) {
+      map[key] = false;
+    } else {
+      return false;
+    }
+  }
+  return true;
```

Comment:
Seems like this checks if `map` is a superset of `set`, instead of equal to it?

Reply:
Standalone that is correct; but checking the length [here](https://github.com/fent/ret.js/blob/5425a32b86523d46b67ca7bfd51251cf0aac4a55/lib/write-set-tokens.ts#L50-L60) beforehand prevents the issue. We have discussed this code section [here](https://github.com/fent/ret.js/pull/25#discussion_r551751830).